File: app/recommendation/similarity/similarity_shard_factory.py
from app.projection.datastore.projection_datastore_shard import ProjectionDatastoreShard
from app.recommendation.similarity.similarity_shard import SimilarityShard


class SimilarityShardFactory:
    '''
    Factory for SimilarityShard. Responsible for building vector space
    and author index, and finding the average vector.
    '''

    def __init__(self, data_shard: ProjectionDatastoreShard):
        self.data_shard = data_shard

    def build(self):
        vectors = []
        author_by_index = dict()
        index = 0
        average_vec = None
        for author, vector in self.data_shard.get_all().items():
            if author == "_average":
                average_vec = vector
            else:
                author_by_index[index] = author
                index += 1
                vectors.append(vector)

        return SimilarityShard(vectors, author_by_index, average_vec)

File: experimental/steven/similarity/similarity_shard_factory.cpp
#include "similarity_shard_factory.h"

#include <vector>

SimilarityShardFactory::SimilarityShardFactory(std::unique_ptr<ProjectionDatastoreShard>&& data_shard) {
    this->data_shard = std::move(data_shard);
}


SimilarityShard SimilarityShardFactory::build() {
    std::vector<std::vector<float>> vectors;
    std::map<int, std::string> author_index;
    std::optional<std::vector<float>> average_vector = std::nullopt;
    std::map<std::string, std::vector<float>> projection = data_shard->getAll();

    int index = 0;
    for (auto projection_iterator = projection.begin(); 
         projection_iterator != projection.end(); projection_iterator++) {
        projection_iterator->first;
        projection_iterator->second;

        if (projection_iterator->first.compare(average_vector_name)) {
            average_vector = projection_iterator->second;
        } else {
            author_index[index++] = projection_iterator->first;
            vectors.push_back(projection_iterator->second);
        }
    }

    return SimilarityShard(vectors, author_index, average_vector);
}